dm panic due to empty column name of unique index

### What did you do?

start dm task

### What did you expect to see?

task runs well

### What did you see instead?

```
DM panic:
[golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75](http://golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75) +0x64
[golang.org/x/sync/errgroup.(*Group).Go.func1()](http://golang.org/x/sync/errgroup.(*Group).Go.func1())
        [github.com/pingcap/tiflow/dm/pkg/checker/worker_pool.go:85](http://github.com/pingcap/tiflow/dm/pkg/checker/worker_pool.go:85) +0xd3
[github.com/pingcap/tiflow/dm/pkg/checker](http://github.com/pingcap/tiflow/dm/pkg/checker).(*WorkerPool[...]).Go.func1()
        [github.com/pingcap/tiflow/dm/pkg/checker/table_structure.go:189](http://github.com/pingcap/tiflow/dm/pkg/checker/table_structure.go:189) +0x7ed
[github.com/pingcap/tiflow/dm/pkg/checker.(*tablesCheckerWorker).handle](http://github.com/pingcap/tiflow/dm/pkg/checker.(*tablesCheckerWorker).handle)(0xc001a70300, {0x493ce28, 0xc0011cc5f0}, 0xc0011cdc70)
        [github.com/pingcap/tiflow/dm/pkg/checker/table_structure.go:338](http://github.com/pingcap/tiflow/dm/pkg/checker/table_structure.go:338) +0x56c
[github.com/pingcap/tiflow/dm/pkg/checker.(*TablesChecker).checkAST](http://github.com/pingcap/tiflow/dm/pkg/checker.(*TablesChecker).checkAST)(0xc0018583e8?, 0xc0011fd450, 0xc0011fdad0, {0x0, 0x0, 0x0})
        [github.com/pingcap/tiflow/dm/pkg/checker/table_structure.go:404](http://github.com/pingcap/tiflow/dm/pkg/checker/table_structure.go:404) +0x2694
[github.com/pingcap/tiflow/dm/pkg/checker.(*TablesChecker).checkTableStructurePair](http://github.com/pingcap/tiflow/dm/pkg/checker.(*TablesChecker).checkTableStructurePair)(0xc001935c48?, 0xc0011fd450, 0xc0011fdad0, {0x0, 0x0, 0xc000098c00?})
        [github.com/pingcap/tiflow/dm/pkg/checker/utils.go:215](http://github.com/pingcap/tiflow/dm/pkg/checker/utils.go:215)
[github.com/pingcap/tiflow/dm/pkg/checker.getPKAndUK(...)](http://github.com/pingcap/tiflow/dm/pkg/checker.getPKAndUK(...))
goroutine 864 [running]:
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x3347634]
panic: runtime error: invalid memory address or nil pointer dereference
```
![img_v2_12cef23c-cbd5-4898-9c04-4d801868385g](https://github.com/pingcap/tiflow/assets/35025882/c55a70cc-c8db-4ec8-a37f-a5275f315af3)


### Versions of the cluster

DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):
```console
master
```
Upstream MySQL/MariaDB server version:
```console
(paste upstream MySQL/MariaDB server version here)
```
Downstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):
```console
(paste TiDB cluster version here)
```
How did you deploy DM: tiup or manually?
```console
(leave TiUP or manually here)
```
Other interesting information (system version, hardware config, etc):
```console
>
>
```

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

--- a/dm/pkg/checker/table_structure_test.go
+++ b/dm/pkg/checker/table_structure_test.go
@@ -984,8 +984,11 @@ func TestExpressionUK(t *testing.T) {
 	result = checker.Check(ctx)
 	require.Equal(t, StateWarning, result.State)
 	require.Len(t, result.Errors, 2)
-	require.Contains(t, result.Errors[0].ShortErr, "upstream has more PK or NOT NULL UK than downstream, index name: uk, columns: [`c2`+1 c]")
-	require.Contains(t, result.Errors[1].ShortErr, "downstream has more PK or NOT NULL UK than upstream, table name: test-table-1, index name: uk, columns: [`c2`+3 c]")
+	// maybe [`c2`+1 c] or [c `c2`+1]
+	require.Contains(t, result.Errors[0].ShortErr, "upstream has more PK or NOT NULL UK than downstream")
+	require.Contains(t, result.Errors[0].ShortErr, "`c2`+1")
+	require.Contains(t, result.Errors[1].ShortErr, "downstream has more PK or NOT NULL UK than upstream")
+	require.Contains(t, result.Errors[1].ShortErr, "`c2`+3")
 	require.NoError(t, mock.ExpectationsWereMet())
 	require.NoError(t, downMock.ExpectationsWereMet())
 }

--- a/dm/pkg/checker/table_structure_test.go
+++ b/dm/pkg/checker/table_structure_test.go
@@ -888,3 +888,104 @@ PRIMARY KEY ("c")
 	mock.ExpectQuery("SHOW CREATE TABLE `test-db`.`test-table-1`").WillReturnRows(createTableRow)
 	return mock
 }
+
+func TestExpressionUK(t *testing.T) {
+	db, mock, err := sqlmock.New()
+	require.NoError(t, err)
+	downDB, downMock, err := sqlmock.New()
+	require.NoError(t, err)
+	ctx := context.Background()
+
+	// test same table structure
+
+	maxConnectionsRow := sqlmock.NewRows([]string{"Variable_name", "Value"}).
+		AddRow("max_connections", "2")
+	mock.ExpectQuery("SHOW VARIABLES LIKE 'max_connections'").WillReturnRows(maxConnectionsRow)
+	sqlModeRow := sqlmock.NewRows([]string{"Variable_name", "Value"}).
+		AddRow("sql_mode", "ANSI_QUOTES")
+	mock.ExpectQuery("SHOW VARIABLES LIKE 'sql_mode'").WillReturnRows(sqlModeRow)
+	createTableRow := sqlmock.NewRows([]string{"Table", "Create Table"}).
+		AddRow("test-table-1", `CREATE TABLE "test-table-1" (
+		  "c" int(11) NOT NULL,
+		  "c2" int(11) NOT NULL,
+		  PRIMARY KEY ("c"),
+		  UNIQUE KEY "uk" (("c2"+1), "c")
+		) ENGINE=InnoDB`)
+	mock.ExpectQuery("SHOW CREATE TABLE `test-db`.`test-table-1`").WillReturnRows(createTableRow)
+	sqlModeRow2 := sqlmock.NewRows([]string{"Variable_name", "Value"}).
+		AddRow("sql_mode", "ANSI_QUOTES")
+	downMock.ExpectQuery("SHOW VARIABLES LIKE 'sql_mode'").WillReturnRows(sqlModeRow2)
+	createTableRow2 := sqlmock.NewRows([]string{"Table", "Create Table"}).
+		AddRow("test-table-1", `CREATE TABLE "test-table-1" (
+		  "c" int(11) NOT NULL,
+		  "c2" int(11) NOT NULL,
+		  PRIMARY KEY ("c"),
+		  UNIQUE KEY "uk" (("c2"+1), "c")
+		) ENGINE=InnoDB`)
+	downMock.ExpectQuery("SHOW CREATE TABLE `test-db`.`test-table-1`").WillReturnRows(createTableRow2)
+
+	checker := NewTablesChecker(
+		map[string]*conn.BaseDB{"test-source": conn.NewBaseDBForTest(db)},
+		conn.NewBaseDBForTest(downDB),
+		map[string]map[filter.Table][]filter.Table{
+			"test-source": {
+				{Schema: "test-db", Name: "test-table-1"}: {
+					{Schema: "test-db", Name: "test-table-1"},
+				},
+			},
+		},
+		nil,
+		1)
+	result := checker.Check(ctx)
+	require.Equal(t, StateSuccess, result.State)
+	require.NoError(t, mock.ExpectationsWereMet())
+	require.NoError(t, downMock.ExpectationsWereMet())
+
+	// test different table structure
+
+	maxConnectionsRow = sqlmock.NewRows([]string{"Variable_name", "Value"}).
+		AddRow("max_connections", "2")
+	mock.ExpectQuery("SHOW VARIABLES LIKE 'max_connections'").WillReturnRows(maxConnectionsRow)
+	sqlModeRow = sqlmock.NewRows([]string{"Variable_name", "Value"}).
+		AddRow("sql_mode", "ANSI_QUOTES")
+	mock.ExpectQuery("SHOW VARIABLES LIKE 'sql_mode'").WillReturnRows(sqlModeRow)
+	createTableRow = sqlmock.NewRows([]string{"Table", "Create Table"}).
+		AddRow("test-table-1", `CREATE TABLE "test-table-1" (
+		  "c" int(11) NOT NULL,
+		  "c2" int(11) NOT NULL,
+		  PRIMARY KEY ("c"),
+		  UNIQUE KEY "uk" (("c2"+1), "c")
+		) ENGINE=InnoDB`)
+	mock.ExpectQuery("SHOW CREATE TABLE `test-db`.`test-table-1`").WillReturnRows(createTableRow)
+	sqlModeRow2 = sqlmock.NewRows([]string{"Variable_name", "Value"}).
+		AddRow("sql_mode", "ANSI_QUOTES")
+	downMock.ExpectQuery("SHOW VARIABLES LIKE 'sql_mode'").WillReturnRows(sqlModeRow2)
+	createTableRow2 = sqlmock.NewRows([]string{"Table", "Create Table"}).
+		AddRow("test-table-1", `CREATE TABLE "test-table-1" (
+		  "c" int(11) NOT NULL,
+		  "c2" int(11) NOT NULL,
+		  PRIMARY KEY ("c"),
+		  UNIQUE KEY "uk" (("c2"+3), "c")
+		) ENGINE=InnoDB`)
+	downMock.ExpectQuery("SHOW CREATE TABLE `test-db`.`test-table-1`").WillReturnRows(createTableRow2)
+
+	checker = NewTablesChecker(
+		map[string]*conn.BaseDB{"test-source": conn.NewBaseDBForTest(db)},
+		conn.NewBaseDBForTest(downDB),
+		map[string]map[filter.Table][]filter.Table{
+			"test-source": {
+				{Schema: "test-db", Name: "test-table-1"}: {
+					{Schema: "test-db", Name: "test-table-1"},
+				},
+			},
+		},
+		nil,
+		1)
+	result = checker.Check(ctx)
+	require.Equal(t, StateWarning, result.State)
+	require.Len(t, result.Errors, 2)
+	require.Contains(t, result.Errors[0].ShortErr, "upstream has more PK or NOT NULL UK than downstream, index name: uk, columns: [`c2`+1 c]")
+	require.Contains(t, result.Errors[1].ShortErr, "downstream has more PK or NOT NULL UK than upstream, table name: test-table-1, index name: uk, columns: [`c2`+3 c]")
+	require.NoError(t, mock.ExpectationsWereMet())
+	require.NoError(t, downMock.ExpectationsWereMet())
+}

--- a/dm/pkg/checker/utils.go
+++ b/dm/pkg/checker/utils.go
@@ -25,7 +25,10 @@ import (
 	"github.com/pingcap/tidb-tools/pkg/utils"
 	"github.com/pingcap/tidb/parser"
 	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/format"
+	"github.com/pingcap/tiflow/dm/pkg/log"
 	"github.com/pingcap/tiflow/dm/pkg/terror"
+	"go.uber.org/zap"
 )
 
 // MySQLVersion represents MySQL version number.
@@ -201,6 +204,8 @@ func getCollation(stmt *ast.CreateTableStmt) string {
 // getPKAndUK returns a map of INDEX_NAME -> set of COLUMN_NAMEs.
 func getPKAndUK(stmt *ast.CreateTableStmt) map[string]map[string]struct{} {
 	ret := make(map[string]map[string]struct{})
+	var sb strings.Builder
+	restoreCtx := format.NewRestoreCtx(format.DefaultRestoreFlags, &sb)
 
 	for _, constraint := range stmt.Constraints {
 		switch constraint.Tp {
@@ -212,7 +217,17 @@ func getPKAndUK(stmt *ast.CreateTableStmt) map[string]map[string]struct{} {
 		case ast.ConstraintUniq, ast.ConstraintUniqKey, ast.ConstraintUniqIndex:
 			ret[constraint.Name] = make(map[string]struct{})
 			for _, key := range constraint.Keys {
-				ret[constraint.Name][key.Column.Name.L] = struct{}{}
+				if key.Column != nil {
+					ret[constraint.Name][key.Column.Name.L] = struct{}{}
+				} else {
+					sb.Reset()
+					err := key.Expr.Restore(restoreCtx)
+					if err != nil {
+						log.L().Warn("failed to restore expression", zap.Error(err))
+						continue
+					}
+					ret[constraint.Name][sb.String()] = struct{}{}
+				}
 			}
 		}
 	}

